### ISSUE TITLE
Model could not converge with normalization features locally 

### ISSUE THREAD

**Issue by fangkuann** [tensorflow/ranking#87]
Hi, 
I trained a model, with dataset pre-normalized by a etl pipeline. This workds fine and the model performed well on the test dataset.
But when I use the transform_fn to normalize features locally, like this
```
     estimator = tf.estimator.Estimator(
         model_fn=tfr.model.make_groupwise_ranking_fn(
             group_score_fn=make_score_fn(),
             group_size=config['group_size'],
             transform_fn=make_transform_fn(config),
             ranking_head=ranking_head),
         config=tf.estimator.RunConfig(
             save_dir,
             save_checkpoints_secs=60 * 30,
             save_checkpoints_steps=50000,
             log_step_count_steps=1000000),
         warm_start_from=warmup_path)
```
and feed raw data without pre-normalization. 
The model could not converge. 
The normalization func is quite simple, just like log(featVal) or (featVal - mean) / std
This phenomenon seems weird, what could be the problem in it?

**Comment by xuanhuiwang** [comment]
It is good to normalize features for DNN in general.

For your problem, I think log(featVal) can be easily plugged in transform_fn and should be the same as you pre-normalize the data. I think you may want to check whether you are **actually** using the log(featVal) in your group_score_fn or not.

(featVal - mean) / std can not be done exactly in transform_fn. You can do batch-level normalization, but this won't be able to be applied during PREDICT time. One the other hand, you can try to add a batch_normalization layer in the group_score_fn to approximate this transformation and this will work during PREDICT time because mean and std are memorized in batch_normalization layer.

**Comment by fangkuann** [comment]
Thanks to @xuanhuiwang for your advice.
I used the tf_debug tool to check the input of the DNN model. I traced multiple features' transformation pipeline and ensured that the features are correctly normalized as I did it in the data preprocessing model. And BN-layer is already added.
I still can not figure out the reason for this weird phenomenon. It may need to dive more into the code and cost more time.  I will update information if I get clear interpretations.

**Comment by fangkuann** [comment]
Hi, @xuanhuiwang . After a little time of debugging. The problem was figured out.
It is because that there are illegal inputs for log() processing, for instance, log(-1), which leads to INF values, and finally makes model training diverge.
The reason that the model works fine when using the ETL pipeline, is that our ETL pipeline finally transforms the log(-xxx) (NaN in scala) to zero values, and it makes models trainable.
Thanks for your advice again !

**Comment by xuanhuiwang** [comment]
@fangkuann Thanks for sharing the findings. This is a good bug and I am glad that you caught it!
